Mathlib/LinearAlgebra/RootSystem/Finite/g2.lean

```diff
@@ -0,0 +1,404 @@
+  calc P.pairing i (threeShortAddLong P) *
+          B.form (threeShortAddLongRoot P) (threeShortAddLongRoot P)
+    _ = 2 * B.form (P.root i) (threeShortAddLongRoot P) := ?_
+    _ = 3 * (2 * B.form (P.root i) (shortRoot P)) + 2 * B.form (P.root i) (longRoot P) := ?_
+    _ = P.pairing i (short P) * B.form (longRoot P) (longRoot P) +
+          P.pairing i (long P) * B.form (longRoot P) (longRoot P) := ?_
+    _ = (P.pairing i (short P) + P.pairing i (long P)) *
+          B.form (threeShortAddLongRoot P) (threeShortAddLongRoot P) := ?_
+  · rw [B.two_mul_apply_root_root]
+  · rw [threeShortAddLongRoot_eq, map_add, mul_add, map_smul, smul_eq_mul]; ring
+  · rw [B.two_mul_apply_root_root, B.two_mul_apply_root_root, long_eq_three_mul_short]; ring
+  · rw [threeShortAddLongRoot_longRoot]; ring
+
+@[simp] lemma pairingIn_threeShortAddTwoLong_left :
+    P.pairingIn ℤ (threeShortAddTwoLong P) i =
+      3 * P.pairingIn ℤ (short P) i + 2 * P.pairingIn ℤ (long P) i := by
+  suffices P.pairing (threeShortAddTwoLong P) i =
+      3 * P.pairing (short P) i + 2 * P.pairing (long P) i from
+    algebraMap_injective ℤ R <| by simpa only [algebraMap_pairingIn, map_add, map_mul, map_ofNat]
+  have : Fintype ι := Fintype.ofFinite ι
+  have B := (P.posRootForm ℤ).toInvariantForm
+  apply mul_right_cancel₀ (B.ne_zero i)
+  rw [← B.two_mul_apply_root_root]
+  simp [threeShortAddTwoLongRoot_eq, mul_add, B.two_mul_apply_root_root, mul_left_comm _ (3 : R)]
+  ring
+
+@[simp] lemma pairingIn_threeShortAddTwoLong_right :
+    P.pairingIn ℤ i (threeShortAddTwoLong P) =
+      P.pairingIn ℤ i (short P) + 2 * P.pairingIn ℤ i (long P) := by
+  suffices P.pairing i (threeShortAddTwoLong P) =
+      P.pairing i (short P) + 2 * P.pairing i (long P) from
+    algebraMap_injective ℤ R <| by simpa only [algebraMap_pairingIn, map_add, map_mul, map_ofNat]
+  have : Fintype ι := Fintype.ofFinite ι
+  have B := (P.posRootForm ℤ).toInvariantForm
+  apply mul_right_cancel₀ (B.ne_zero <| threeShortAddTwoLong P)
+  calc P.pairing i (threeShortAddTwoLong P) *
+          B.form (threeShortAddTwoLongRoot P) (threeShortAddTwoLongRoot P)
+    _ = 2 * B.form (P.root i) (threeShortAddTwoLongRoot P) := ?_
+    _ = 3 * (2 * B.form (P.root i) (shortRoot P)) + 2 * (2 * B.form (P.root i) (longRoot P)) := ?_
+    _ = P.pairing i (short P) * B.form (longRoot P) (longRoot P) +
+          2 * P.pairing i (long P) * B.form (longRoot P) (longRoot P) := ?_
+    _ = (P.pairing i (short P) + 2 * P.pairing i (long P)) *
+          B.form (threeShortAddTwoLongRoot P) (threeShortAddTwoLongRoot P) := ?_
+  · rw [B.two_mul_apply_root_root]
+  · simp only [threeShortAddTwoLongRoot_eq, map_add, mul_add, map_smul, smul_eq_mul]; ring
+  · rw [B.two_mul_apply_root_root, B.two_mul_apply_root_root, long_eq_three_mul_short]; ring
+  · rw [threeShortAddTwoLongRoot_longRoot]; ring
+
+end Pairing
+
+private lemma isOrthogonal_short_and_long_aux {a b c d e f a' b' c' d' e' f' : ℤ} {S : Set (ℤ × ℤ)}
```

Comment:
I think this lemma is a good example of the sort of thing that @hrmacbeth mentions here: https://arxiv.org/abs/2405.04699